user-interface/src/main/java/life/qbic/datamanager/views/projects/purchase/UploadPurchaseDialog.java

```diff
@@ -0,0 +1,193 @@
+      toggleFileSectionIfEmpty();
+    });
+
+    // Synchronise the Vaadin upload component with the purchase list display
+    // When a file is removed  from the upload component, we also want to remove it properly from memory
+    // and from any additional display
+    upload.getElement().addEventListener("file-remove", this::processClientFileRemoveEvent)
+        .addEventData(VAADIN_FILENAME_EVENT);
+
+    // Put the elements together
+    add(titleBox, upload, restrictions, disclaimer, uploadedPurchaseItems);
+    addClassName("purchase-item-upload");
+    confirmButton.setText("Save");
+
+    // Init the visibility rendering once
+    toggleFileSectionIfEmpty();
+  }
+
+  @Override
+  protected void onConfirmClicked(ClickEvent<Button> clickEvent) {
+    fireConfirmEvent();
+  }
+
+  @Override
+  protected void onCancelClicked(ClickEvent<Button> clickEvent) {
+    this.close();
+  }
+
+  private void fireConfirmEvent() {
+    var confirmEvent = new ConfirmEvent<>(this, true);
+    confirmListeners.forEach(listener -> listener.onComponentEvent(confirmEvent));
+  }
+
+  private void processClientFileRemoveEvent(DomEvent event) {
+    JsonObject jsonObject = event.getEventData();
+    var fileName = jsonObject.getString(VAADIN_FILENAME_EVENT);
+    removeFile(fileName);
+  }
+
+  private void removeFile(String fileName) {
+    removeFileFromBuffer(fileName);
+    removeFileFromDisplay(fileName);
+    toggleFileSectionIfEmpty();
+  }
+
+  private void toggleFileSectionIfEmpty() {
+    if (uploadedPurchaseItems.getChildren().toList().isEmpty()) {
+      uploadedPurchaseItems.addClassName("hidden");
+      disclaimer.addClassName("hidden");
+    } else {
+      uploadedPurchaseItems.removeClassName("hidden");
+      disclaimer.removeClassName("hidden");
+    }
+  }
+
+  private void removeFileFromBuffer(String fileName) {
+    multiFileMemoryBuffer.remove(fileName);
+  }
+
+  private void removeFileFromDisplay(String fileName) {
+    var result = purchaseItemsCache.stream().filter(purchaseItem -> purchaseItem.fileName().equals(
+        fileName)).findAny();
+    result.ifPresent(purchaseItem -> {
+      purchaseItemsCache.remove(purchaseItem);
+      uploadedPurchaseItems.remove(purchaseItem);
+    });
+  }
+
+  public List<OfferDTO> purchaseItems() {
+    return purchaseItemsCache.stream().map(this::convertToOffer).toList();
+  }
+
+  private OfferDTO convertToOffer(PurchaseItem purchaseItem) {
+    try {
+      var fileName = purchaseItem.fileName();
+      var content = multiFileMemoryBuffer.inputStream(fileName)
+          .orElse(new ByteArrayInputStream(new byte[]{})).readAllBytes();
+      return new OfferDTO(purchaseItem.isSigned(), fileName, content);
+    } catch (IOException e) {
+      log.error("Failed to read uploaded purchase item content", e);
+      throw new ApplicationException("Failed to read offer content");
+    }
+  }
+
+  public void subscribe(ComponentEventListener<ConfirmEvent<UploadPurchaseDialog>> listener) {
```

Comment:
please rename to `addConfirmListener` to match our general listener method style.